### PR TITLE
Csr check git installation and show sign-in button

### PR DESCRIPTION
**Pull request by Deren-Liao**
*  Add a UI state, if git is not installed, show the UI that asks to install
*  Remove "connect" button, replace it with "sign in" button.   And remove disconnect button.  
   This way, if user is signed in,  it shows repo list.  If user is not singed in, let user add account by clicking sign in button.
*  Show error notification if git set config, git set credential fails ---  this should happen very very low chance because they are simply local file change using git command.

### CODE REVIEW

File: GoogleCloudExtension/GoogleCloudExtension/CloudSourceRepositories/CsrSectionControlViewModel.cs [hunk 1]
```diff
@@ -34,14 +34,16 @@ public class CsrSectionControlViewModel : ViewModelBase, ISectionViewModel
...
         /// Sometimes, the view and view model is recreated by Team Explorer.
         /// This is to preserve the states when a new user control is created.
-        private static bool s_isConnected = false;
         private static string s_currentAccount;
         private static bool s_gitInited;
 
+        private ITeamExplorerUtils _teamExplorerService;
         private readonly CsrReposContent _reposContent = new CsrReposContent();
         private readonly CsrUnconnectedContent _unconnectedContent = new CsrUnconnectedContent();
+        private readonly CsrGitSetupWarningContent _gitSetupContent = new CsrGitSetupWarningContent();
```
Comment (ILMTitan) [inline GoogleCloudExtension/GoogleCloudExtension/CloudSourceRepositories/CsrSectionControlViewModel.cs]: I am worried that all of these UI classes in the ViewModel will make testing the ViewModel difficult in the future.

File: GoogleCloudExtension/GoogleCloudExtension/CloudSourceRepositories/CsrSectionControlViewModel.cs [hunk 2]
```diff
@@ -63,36 +65,51 @@ public CsrSectionControlViewModel()
...
             {
-                return;
+                ManageAccountsWindow.PromptUser();
             }
+            Refresh();
+        }
+
+        /// <summary>
+        /// Continues to show unconnected or connected view,
+        /// after it checks git is installed.
```
Comment (ILMTitan) [inline GoogleCloudExtension/GoogleCloudExtension/CloudSourceRepositories/CsrSectionControlViewModel.cs]: *Continue on by showing the ...

File: GoogleCloudExtension/GoogleCloudExtension/CloudSourceRepositories/CsrSectionControlViewModel.cs [hunk 3]
```diff
@@ -63,36 +65,51 @@ public CsrSectionControlViewModel()
...
-            if (CredentialsStore.Default.CurrentAccount != null)
+            else
             {
                 Content = _reposContent;
-                s_isConnected = true;
-                Refresh();
+                if (s_currentAccount != CredentialsStore.Default.CurrentAccount?.AccountName)
+                {
+                    SetGitCredential(_teamExplorerService);
+                    // Continue nevertheless SetGitCredential succeeds or not.
```
Comment (ILMTitan) [inline GoogleCloudExtension/GoogleCloudExtension/CloudSourceRepositories/CsrSectionControlViewModel.cs]: *Continue regardless of whether ...

File: GoogleCloudExtension/GoogleCloudExtension/CloudSourceRepositories/CsrSectionControlViewModel.cs [hunk 4]
```diff
@@ -165,42 +180,59 @@ void ISectionViewModel.Cleanup()
...
             {
                 return true;
             }
+
+            return s_gitInited = (await SetUseHttpPath(teamExplorer)) && SetGitCredential(teamExplorer);
+        }
+
+        private static async Task<bool> SetUseHttpPath(ITeamExplorerUtils teamExplorer)
+        {
+            bool ret = false;
```
Comment (ILMTitan) [inline GoogleCloudExtension/GoogleCloudExtension/CloudSourceRepositories/CsrSectionControlViewModel.cs]: R#: Initializer is redundant.

File: GoogleCloudExtension/GoogleCloudExtension/Utils/ValidateGitDependencyHelper.cs [hunk 5]
```diff
@@ -15,6 +15,7 @@
 using GoogleCloudExtension.GitUtils;
 using GoogleCloudExtension.LinkPrompt;
 using System;
+using System.Threading;
```
Comment (ILMTitan) [inline GoogleCloudExtension/GoogleCloudExtension/Utils/ValidateGitDependencyHelper.cs]: R#: Unnecessary `using` directive.

File: GoogleCloudExtension/GoogleCloudExtension/Resources.resx [hunk 6]
```diff
@@ -2643,10 +2643,6 @@
     <value>Clone or Create</value>
     <comment>The clone or create repo link button caption</comment>
   </data>
-  <data name="CsrConnectLinkButtonCaption" xml:space="preserve">
-    <value>Connect...</value>
-    <comment>The connect link button caption</comment>
-  </data>
```
Comment (ILMTitan) [inline GoogleCloudExtension/GoogleCloudExtension/Resources.resx]: Why is this being removed? It is still used in *CsrUnconnectedContent.xaml*.

File: GoogleCloudExtension/GoogleCloudExtension/CloudSourceRepositories/CsrSectionControlViewModel.cs [hunk 7]
```diff
@@ -63,36 +65,51 @@ public CsrSectionControlViewModel()
...
-            bool inited = await GitConfig();
-            if (!inited)
-            {
-                return;
-            }
-            if (s_isConnected)
+            if (CredentialsStore.Default.CurrentAccount == null)
             {
-                return;
+                ManageAccountsWindow.PromptUser();
```
Comment (Deren-Liao) [inline GoogleCloudExtension/GoogleCloudExtension/CloudSourceRepositories/CsrSectionControlViewModel.cs]: It goes to ShowUnconnectedView at Refresh().

File: GoogleCloudExtension/GoogleCloudExtension/CloudSourceRepositories/CsrSectionControlViewModel.cs (hunk 2 above)
Comment (Deren-Liao) [inline GoogleCloudExtension/GoogleCloudExtension/CloudSourceRepositories/CsrSectionControlViewModel.cs]: Done

File: GoogleCloudExtension/GoogleCloudExtension/CloudSourceRepositories/CsrGitSetupWarningViewModel.cs [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+        {
+            _parent = parent;
+            InstallGitCommand = new ProtectedCommand(
+                () => Process.Start(ValidateGitDependencyHelper.GitInstallationLink));
+            VerifyCommand = new ProtectedAsyncCommand(VerifyInstallation);
+        }
+
+        /// <summary>
+        /// Check if Git for Windows dependency is installed properly.
+        /// Set s_error so that the error shows
```
Comment (Deren-Liao) [inline GoogleCloudExtension/GoogleCloudExtension/CloudSourceRepositories/CsrGitSetupWarningViewModel.cs]: yes. Done

File: GoogleCloudExtension/GoogleCloudExtension/CloudSourceRepositories/CsrGitSetupWarningViewModel.cs [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+
+        /// <summary>
+        /// Indicates if the git installation has been verified installed.
+        /// </summary>
+        public static bool GitInstallationVerified { get; private set; }
+
+        /// <summary>
+        /// Enalbe or disable the control
+        /// </summary>
+        public bool IsEnable
```
Comment (Deren-Liao) [inline GoogleCloudExtension/GoogleCloudExtension/CloudSourceRepositories/CsrGitSetupWarningViewModel.cs]: Done

File: GoogleCloudExtension/GoogleCloudExtension/CloudSourceRepositories/CsrGitSetupWarningViewModel.cs [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+            if (String.IsNullOrWhiteSpace(GitRepository.GetGitPath()))
+            {
+                ErrorMessage = Resources.GitUtilsMissingGitErrorTitle;
+                return;
+            }
+            try
+            {
+                await GitRepository.GitCredentialManagerInstalled();
+            }
+            catch (GitCommandException)
```
Comment (Deren-Liao) [inline GoogleCloudExtension/GoogleCloudExtension/CloudSourceRepositories/CsrGitSetupWarningViewModel.cs]: Let me change the IsGitCredentialManagerInstalledAsync to return bool.

File: GoogleCloudExtension/GoogleCloudExtension/Resources.resx (hunk 6 above)
Comment (Deren-Liao) [inline GoogleCloudExtension/GoogleCloudExtension/Resources.resx]: Changed to CsrSignInLinkButtonCaption

File: GoogleCloudExtension/GoogleCloudExtension/CloudSourceRepositories/CsrGitSetupWarningViewModel.cs [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+            {
+                ErrorMessage = Resources.GitUtilsGitCredentialManagerNotInstalledMessage;
+                return;
+            }
+
+            ErrorMessage = null;
+            GitInstallationVerified = true;
+        }
+
+        private async Task VerifyInstallation()
```
Comment (Deren-Liao) [inline GoogleCloudExtension/GoogleCloudExtension/CloudSourceRepositories/CsrGitSetupWarningViewModel.cs]: It's true. Then it reads odd that it calls to _parent and comes back to this.CheckInstallation ...  Very confusion.
I merged CheckGitInstallation and VerifyInstallation.

File: GoogleCloudExtension/GoogleCloudExtension/CloudSourceRepositories/CsrSectionControlViewModel.cs (hunk 1 above)
Comment (Deren-Liao) [inline GoogleCloudExtension/GoogleCloudExtension/CloudSourceRepositories/CsrSectionControlViewModel.cs]: I'll modify them later if unit test needs such modification.

File: GoogleCloudExtension/GoogleCloudExtension/CloudSourceRepositories/CsrSectionControlViewModel.cs (hunk 3 above)
Comment (Deren-Liao) [inline GoogleCloudExtension/GoogleCloudExtension/CloudSourceRepositories/CsrSectionControlViewModel.cs]: Done

File: GoogleCloudExtension/GoogleCloudExtension/CloudSourceRepositories/CsrSectionControlViewModel.cs (hunk 4 above)
Comment (Deren-Liao) [inline GoogleCloudExtension/GoogleCloudExtension/CloudSourceRepositories/CsrSectionControlViewModel.cs]: Done

File: GoogleCloudExtension/GoogleCloudExtension/Utils/ValidateGitDependencyHelper.cs (hunk 5 above)
Comment (Deren-Liao) [inline GoogleCloudExtension/GoogleCloudExtension/Utils/ValidateGitDependencyHelper.cs]: Done

File: GoogleCloudExtension/GoogleCloudExtension/CloudSourceRepositories/CsrGitSetupWarningViewModel.cs (hunk 10 above)
Comment (Deren-Liao) [inline GoogleCloudExtension/GoogleCloudExtension/CloudSourceRepositories/CsrGitSetupWarningViewModel.cs]: Done

File: GoogleCloudExtension/GoogleCloudExtension/CloudSourceRepositories/CsrGitSetupWarningViewModel.cs (hunk 8 above)
Comment (Deren-Liao) [inline GoogleCloudExtension/GoogleCloudExtension/CloudSourceRepositories/CsrGitSetupWarningViewModel.cs]: Done

File: GoogleCloudExtension/GoogleCloudExtension/CloudSourceRepositories/CsrGitSetupWarningViewModel.cs (hunk 9 above)
Comment (Deren-Liao) [inline GoogleCloudExtension/GoogleCloudExtension/CloudSourceRepositories/CsrGitSetupWarningViewModel.cs]: Done

File: GoogleCloudExtension/GoogleCloudExtension/CloudSourceRepositories/CsrGitSetupWarningViewModel.cs [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+        private bool _isEnabled = false;
+        private string _errorMessage;
+
+        /// <summary>
+        /// Indicates if the git installation has been verified installed.
+        /// </summary>
+        public static bool GitInstallationVerified { get; private set; }
+
+        /// <summary>
+        /// Enalbe or disable the control
```
Comment (Deren-Liao) [inline GoogleCloudExtension/GoogleCloudExtension/CloudSourceRepositories/CsrGitSetupWarningViewModel.cs]: Done

File: GoogleCloudExtension/GoogleCloudExtension.GitUtils/GitRepository.cs [hunk 13]
```diff
@@ -120,6 +120,12 @@ public Task<List<string>> GetRevisionFileAsync(string sha, string relativePath)
             RunGitCommandAsync(command, Root, throwOnError: false);
 
         /// <summary>
+        /// Verifies git credential manager is installed propoerly.
+        /// </summary>
+        public static Task GitCredentialManagerInstalled() =>
```
Comment (Deren-Liao) [inline GoogleCloudExtension/GoogleCloudExtension.GitUtils/GitRepository.cs]: Done

File: GoogleCloudExtension/GoogleCloudExtension.GitUtils/GitRepository.cs [hunk 14]
```diff
@@ -120,6 +120,12 @@ public Task<List<string>> GetRevisionFileAsync(string sha, string relativePath)
             RunGitCommandAsync(command, Root, throwOnError: false);
 
         /// <summary>
+        /// Verifies git credential manager is installed propoerly.
```
Comment (Deren-Liao) [inline GoogleCloudExtension/GoogleCloudExtension.GitUtils/GitRepository.cs]: Done

File: GoogleCloudExtension/GoogleCloudExtension/CloudSourceRepositories/CsrGitSetupWarningContent.xaml [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+            HorizontalAlignment="Left" 
+            VerticalAlignment="Top"        
+            Stretch="None"
+            Source="{mp:ImageResource Theming/Resources/GCP_logo_horizontal.png}" />
+
+        <Grid>
+            <Grid.ColumnDefinitions>
+                <ColumnDefinition Width="Auto" />
+                <ColumnDefinition Width="*" />
+            </Grid.ColumnDefinitions>
```
Comment (Deren-Liao) [inline GoogleCloudExtension/GoogleCloudExtension/CloudSourceRepositories/CsrGitSetupWarningContent.xaml]: Done

**Comment by Deren-Liao** [comment]
Updated, please take a look again.

**Comment by Deren-Liao** [comment]
Thanks!

**Pull request merged**